Optimize native forms v2

bennsimon opened the pull request:
https://github.com/OpenSRP/opensrp-client-anc/pull/511
- Make number of contact schedules shown on summary configurable
- Add `contact_no` to globals for quick check, when its opened from register Activity
- Fix null GA
- Fix incorrect count of profile contact container relevant required fields

- Optimized native forms
- Support use of boolean `is_first_contact` as due check strategy
- Make `ProfileActivity.class` configurable

Commit:
make number of contacts schedule displayed on summary configurable

add contact no to globals when quick check is opened from register

Commit:
fix null GA

fix required field counts

Commit:
updated code to tests native forms performance improvement

Commit:
code cleanup

Commit:
fix gibberish text and crash in ct

Commit:
update to local snapshot

Commit:
update code with latest optimization code

Commit:
update popStack

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/repository/PatientRepository.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/util/AncMetadata.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
add contact sequencing logic

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/util/AncMetadata.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/repository/PatientRepository.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
remove due check strategy config

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
make anc profile activity configurable

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
update snapshot version

Commit:
Merge branch 'develop' of github.com:OpenSRP/opensrp-client-anc into optimize-native-forms-v2

Commit:
remove unneeded column

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
update null check for event details

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
fix next contact update

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
update/fix tests

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  5
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/BaseActivityUnitTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

keymane commented on opensrp-anc/src/test/java/org/smartregister/anc/library/activity/BaseHomeRegisterActivityTest.java:
```diff
@@ -18,6 +18,7 @@
 import org.robolectric.Robolectric;
 import org.robolectric.RuntimeEnvironment;
 import org.robolectric.android.controller.ActivityController;
+import org.robolectric.util.ReflectionHelpers;
```
Codacy found an issue: [Avoid unused imports such as 'org.robolectric.util.ReflectionHelpers'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791)

keymane replied:
Codacy found an issue: [Avoid unused imports such as 'org.robolectric.util.ReflectionHelpers'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791)

keymane commented on opensrp-anc/src/test/java/org/smartregister/anc/library/interactor/RegisterInteractorTest.java:
```diff
@@ -62,8 +67,12 @@
     @Captor
     private ArgumentCaptor<Long> longArgumentCaptor;
 
+    @Mock
```
Codacy found an issue: [Use explicit scoping instead of the default package private level](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791)

keymane replied:
Codacy found an issue: [Use explicit scoping instead of the default package private level](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791)

keymane commented on opensrp-anc/src/test/java/org/smartregister/anc/library/activity/BaseHomeRegisterActivityTest.java:
```diff
@@ -32,6 +33,7 @@
 import org.smartregister.anc.library.util.ConstantsUtils;
 import org.smartregister.anc.library.util.DBConstantsUtils;
 import org.smartregister.commonregistry.CommonPersonObjectClient;
+import org.smartregister.configurableviews.ConfigurableViewsLibrary;
```
Codacy found an issue: [Avoid unused imports such as 'org.smartregister.configurableviews.ConfigurableViewsLibrary'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791)

keymane replied:
Codacy found an issue: [Avoid unused imports such as 'org.smartregister.configurableviews.ConfigurableViewsLibrary'](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791)

Commit:
added tests

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  5
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/BaseActivityUnitTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
fix tests

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/BaseActivityUnitTest.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  5
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

dubdabasoduba commented:
@bennsimon is this ready for review?

dubdabasoduba commented on gradle.properties:
```diff
@@ -8,10 +8,12 @@
 
 ## PUBLISHING VARS
-VERSION_NAME=2.0.1-SNAPSHOT
+VERSION_NAME=2.0.3.13-LOCAL-SNAPSHOT
```
Change this version name

dubdabasoduba commented on opensrp-anc/build.gradle:
```diff
@@ -143,7 +143,7 @@ dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
     implementation 'com.android.support:appcompat-v7:28.0.0'
-    implementation('org.smartregister:opensrp-client-native-form:1.7.28-SNAPSHOT@aar') {
+    implementation('org.smartregister:opensrp-client-native-form:1.9.2-r117-OPTIMIZED-LOCAL-SNAPSHOT@aar') {
```
Update to the correct version name

dubdabasoduba commented on opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java:
```diff
@@ -90,6 +93,12 @@ private void initializeMainContactContainers() {
             requiredFieldsMap.put(ConstantsUtils.JsonFormUtils.ANC_TEST_TASKS_ENCOUNTER_TYPE, 0);
 
+            if (StringUtils.isNotBlank(formInvalidFields) && contactNo > 1 && !PatientRepository.isFirstVisit(baseEntityId)) {
+                String[] pair = formInvalidFields.split(":");
```
Should check the size of this string array. Otherwise, you will get an out of bounds exception

dubdabasoduba commented on opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java:
```diff
@@ -318,7 +328,7 @@ private void processRequiredStepsField(JSONObject object) throws Exception {
             while (keys.hasNext()) {
                 String key = keys.next();
-                if (key.startsWith(RuleConstant.STEP)) {
+                if (key.startsWith(RuleConstant.STEP) && !object.getJSONObject(key).has("skipped")) {
```
Move the `skipped` string to constants

dubdabasoduba commented on opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java:
```diff
@@ -761,4 +775,11 @@ protected void onPause() {
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (resultCode == RESULT_OK && data != null) {
+            formInvalidFields = data.getStringExtra("formInvalidFields");
```
Move this string to constants

dubdabasoduba commented on reference-app/build.gradle:
```diff
@@ -130,7 +130,7 @@ android {
 
         debug {
-            resValue "string", 'opensrp_url', '"https://anc-stage.smartregister.org/opensrp/"'
+            resValue "string", 'opensrp_url', '"https://opensrp-jembi-eregister.smartregister.org/opensrp/"' //TODO should be changed to anc staging url
```
Update this to use the ANC reference app URL

dubdabasoduba commented on reference-app/build.gradle:
```diff
@@ -221,17 +221,17 @@ tasks.withType(Test) {
-        exclude group: 'org.yaml', module: 'snakeyaml'
-        exclude group: 'io.ona.rdt-capture', module: 'lib'
-    }
+//    implementation('org.smartregister:opensrp-client-native-form:1.7.30-OPTIMIZE_FORMS-SNAPSHOT@aar') {
+//        transitive = true
+//        exclude group: 'com.android.support', module: 'recyclerview-v7'
+//        exclude group: 'com.android.support', module: 'appcompat-v7'
+//        exclude group: 'com.android.support', module: 'cardview-v7'
+//        exclude group: 'com.android.support', module: 'support-media-compat'
+//        exclude group: 'com.android.support', module: 'support-v4'
+//        exclude group: 'com.android.support', module: 'design'
+//        exclude group: 'org.yaml', module: 'snakeyaml'
+//        exclude group: 'io.ona.rdt-capture', module: 'lib'
+//    }
```
Why have you disabled this?

Commit:
update opensrp url to anc

add check got contact summar model to avoid duplicates

keymane commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-anc/src/main/java/org/smartregister/anc/library/model/ContactSummaryModel.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/MainContactActivity.java  1
- opensrp-anc/src/test/java/org/smartregister/anc/library/activity/BaseActivityUnitTest.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/task/LoadContactSummaryDataTask.java  5
- opensrp-anc/src/main/java/org/smartregister/anc/library/fragment/ProfileContactsFragment.java  1
- opensrp-anc/src/main/java/org/smartregister/anc/library/activity/ContactSummarySendActivity.java  3
- opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ContactSummaryPresenter.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-anc/pullRequest?prid=5624791&bid=18425333)

Commit:
add anc maternity transfer logic

Commit:
update anc maternity transfer logic

Commit:
update logic sor AncMaternityTransferProcessor

Commit:
update anc death processing & transfer processor

Commit:
code cleanup

Commit:
fix tests

Commit:
add tests

Commit:
fix codacy issue

Commit:
fix tests

Commit:
remove update of key in repeating grp

Commit:
add test

Commit:
migrate to latest core

Commit:
add 41 weeks limit

dubdabasoduba committed:
Merge pull request #540 from opensrp/migrate-to-oauth2

Migrate to oauth2

ekigamba committed:
Fix next page fragment being transacted from diskIO

- Update dependencies to com.github.lecho:hellocharts-android:1.5.8 and id.zelory:compressor:2.1.1
- Update 2.0.3.40-LOCAL-SNAPSHOT
